fish-rust/src/signal.rs

```diff
@@ -65,3 +67,272 @@ pub fn signal_get_desc(sig: i32) -> &'static wstr {
+            Signal::SIGSTKFLT => "SIGSTKFLT",
+            #[cfg(target_os = "linux")]
+            Signal::SIGPWR => "SIGPWR",
+            Signal(_) => Self::UNKNOWN_SIG_NAME,
+        }
+    }
+
+    pub const fn desc(&self) -> &'static str {
+        match *self {
+            Signal::SIGHUP => "Terminal hung up",
+            Signal::SIGINT => "Quit request from job control (^C)",
+            Signal::SIGQUIT => "Quit request from job control with core dump (^\\)",
+            Signal::SIGILL => "Illegal instruction",
+            Signal::SIGTRAP => "Trace or breakpoint trap",
+            Signal::SIGABRT => "Abort",
+            #[cfg(any(target_os = "freebsd", target_os = "macos"))]
+            Signal::SIGEMT => "Emulator trap",
+            Signal::SIGFPE => "Floating point exception",
+            Signal::SIGKILL => "Forced quit",
+            Signal::SIGBUS => "Misaligned address error",
+            Signal::SIGSEGV => "Address boundary error",
+            Signal::SIGSYS => "Bad system call",
+            Signal::SIGPIPE => "Broken pipe",
+            Signal::SIGALRM => "Timer expired",
+            Signal::SIGTERM => "Polite quit request",
+            Signal::SIGURG => "Urgent socket condition",
+            Signal::SIGSTOP => "Forced stop",
+            Signal::SIGTSTP => "Stop request from job control (^Z)",
+            Signal::SIGCONT => "Continue previously stopped process",
+            Signal::SIGCHLD => "Child process status changed",
+            Signal::SIGTTIN => "Stop from terminal input",
+            Signal::SIGTTOU => "Stop from terminal output",
+            Signal::SIGIO => "I/O on asynchronous file descriptior is possible",
+            Signal::SIGXCPU => "CPU time limit exceeded",
+            Signal::SIGXFSZ => "File size limit exceeded",
+            Signal::SIGVTALRM => "Virtual timer expired",
+            Signal::SIGPROF => "Profiling timer expired",
+            Signal::SIGWINCH => "Window size change",
+            #[cfg(any(target_os = "freebsd", target_os = "macos"))]
+            Signal::SIGINFO => "Information request",
+            Signal::SIGUSR1 => "User-defined signal 1",
+            Signal::SIGUSR2 => "User-defined signal 2",
+            #[cfg(any(target_os = "freebsd"))]
+            Signal::SIGTHR => "Thread interrupt",
+            #[cfg(any(target_os = "freebsd"))]
+            Signal::SIGLIBRT => "Real-time library interrupt",
+            #[cfg(target_os = "linux")]
+            Signal::SIGSTKFLT => "Stack fault",
+            #[cfg(target_os = "linux")]
+            Signal::SIGPWR => "Power failure",
+            Signal(_) => "Unknown",
+        }
+    }
+
+    pub const fn code(&self) -> i32 {
+        // Safety: Types share size and alignment but one just has additional semantic constraints
+        (unsafe { *(&self.0 as *const _ as *const u8) }) as i32
+    }
+
+    pub const fn new(sig: NonZeroU8) -> Self {
+        Signal(sig)
+    }
+
+    pub fn try_new(sig: i32) -> Option<Self> {
```

Comment:
I think we should just have a single function which either accepts a NonZeroI32, or an i32 and asserts that it's non-zero. I don't see a reason to make a fallible version here.

Reply:
It was to avoid asserting/panicking but keep the option of passing in an i32. If you think it's OK to panic if we get a zero, that makes things easier.